![vaccine](images/vaccine.png)

# H1N1 and Seasonal Flu Vaccines

**Author**: Ian Butler

***

## Overview

A one-paragraph overview of the project, including the business problem, data, methods, results and recommendations.

***

## Business Problem

Summary of the business problem we are trying to solve, and the data questions that we plan to answer to solve them.

***

Questions to consider:

* What are the business's pain points related to ths project?
    * The business's pain points related to this project are these:
        * Primary: can we predict whether people got H1N1 vaccines using data collected in the National 2009 H1N1 Flu Survey?
        * Secondary: can we infer which variables of this survey are most causing a change in vaccination status for H1N1?
        * Tertiary: can we provide actionable recommendations to improve the rate of innoculation against H1N1?
* How did we pick the data analysis question(s) that we did?
    * With respect to each of the above questions, we picked the questions as a result of the following thought process:
        * Ultimately, what we would like to be able to do is increase the population's rate of innoculation against highly infectious respiratory diseases. In order to know what would have to be done to take action in this way, we have to understand what variables are most causing a change in vaccination status for individuals, as well as which of those variables we can actually control. Finally, we can confirm whether or not our understanding of these relationships can be used to accurately predict vaccination status for individuals.
* Why are these questions important from a business perspective?
    * These questions are important from a business perspective because they allow public health officials to make informed decisions about what actions they can take to attempt to improve the resiliency of the population against highly infectious respiratory diseases.

***

## Data Understanding

Describe the data being used for this project.

***

Questions to consider:

* Where did the data come from?
    * The data comes from the National 2009 H1N1 Flu Survey.
* How do they relate to the data analysis questions?
    * The National 2009 H1N1 Flu Survery relates to the data analysis questions in that this survey was specifically designed to develop an understanding of how people’s backgrounds, opinions, and health behaviors are related to their personal vaccination patterns.
* What do the data represent?
    * Each entry in the data represents the survey responses of one individual.
* Who is in the sample?
    * The sample is comprised of individuals from various psychographics and demographics across the United States.
* What variable are includeed?
    * The variables included are as follows:
    * For all binary variables: 0 = No; 1 = Yes.
        * respondent_id
            * a unique and random identifier
        * h1n1_concern - Level of concern about the H1N1 flu.
            * 0 = Not at all concerned;
            * 1 = Not very concerned;
            * 2 = Somewhat concerned;
            * 3 = Very concerned.
        * h1n1_knowledge - Level of knowledge about the H1N1 flu.
            * 0 = No knowledge;
            * 1 = A little knowledge;
            * 2 = A lot of knowledge.
        * behavioral_antiviral_meds - Has taken antiviral medications.
            * (binary)
        * behavioral_avoidance - Has avoided close contact with other with flu-like symptoms.
            * (binary)
        * behavioral_face_mask - Has bought a face mask.
            * (binary)
        * behavioral_wash_hands - Has frequently washed hands or used hand sanitizer.
            * (binary)
        * behavioral_large_gatherings - Has reduced time at large gatherings.
            * (binary)
        * behavioral_outside_home - Has reduced contact with people outside of own household.
            * (binary)
        * behavioral_touch_face - Has avoided touching eyes, nose, or mouth.
            * (binary)
        * doctor_recc_h1n1 - H1N1 flu vaccine was recommended by doctor.
            * (binary)
        * doctor_recc_seasonal - Seasonal flu vaccine was recommended by doctor.
            * (binary)
        * chronic_med_condition - Has any of the following chronic medical conditions: asthma or an other lung condition, diabetes, a heart condition, a kidney condition, sickle cell anemia or other anemia, a neurological or neuromuscular condition, a liver condition, or a weakened immune system caused by a chronic illness or by medicines taken for a chronic illness.
            * (binary)
        * child_under_6_months - Has regular close contact with a child under the age of six months.
            * (binary)
        * health_worker - Is a healthcare worker.
            * (binary)
        * health_insurance - Has health insurance.
            * (binary)
        * opinion_h1n1_vacc_effective - Respondent's opinion about H1N1 vaccine effectiveness.
            * 1 = Not at all effective;
            * 2 = Not very effective;
            * 3 = Don't know;
            * 4 = Somehwat effective;
            * 5 = Very effective.
        * opinion_h1n1_risk - Respondent's opinion about risk of getting sick with H1N1 flu without vaccine.
            * 1 = Very low;
            * 2 = Somewhat low;
            * 3 = Don't know;
            * 4 = Somewhat high;
            * 5 = Very high.
        * opinion_h1n1_sick_from_vacc - Respondent's worry about getting sick from taking H1N1 vaccine.
            * 1 = Not at all worried;
            * 2 = Not very worried;
            * 3 = Don't know;
            * 4 = Somewhat worried;
            * 5 = Very worried.
        * opinion_seas_vacc_effective - Respondent's opinion about seasonal flu vaccine effectiveness.
            * 1 = Not at all effective;
            * 2 = Not very effective;
            * 3 = Don't know;
            * 4 = Somehwat effective;
            * 5 = Very effective.
        * opinion_seas_risk - Respondent's opinion about risk of getting sick with seasonal flu without vaccine.
            * 1 = Very low;
            * 2 = Somewhat low;
            * 3 = Don't know;
            * 4 = Somewhat high;
            * 5 = Very high.
        * opinion_seas_sick_from_vacc - Respondent's worry about getting sick from taking seasonal flu vaccine.
            * 1 = Not at all worried;
            * 2 = Not very worried;
            * 3 = Don't know;
            * 4 = Somewhat worried;
            * 5 = Very worried.
        * age_group - Age group of respondent.
        * education - Self-reported education level.
        * race - Race of respondent.
        * sex - Sex of respondent.
        * income_poverty - Household annual income of respondent with respect to 2008 Census poverty thresholds.
        * marital_status - Marital status of respondent.
        * rent_or_own - Housing situation of respondent.
        * employment_status - Employment status of respondent.
        * hhs_geo_region - Respondent's residence using a 10-region geographic classification defined by the U.S. Dept. of Health and Human Services. Values are represented as short random character strings.
        * census_msa - Respondent's residence within metropolitan statistical areas (MSA) as defined by the U.S. Census.
        * household_adults - Number of *other* adults in household, top-coded to 3.
        * household_children - Number of children in household, top-coded to 3.
        * employment_industry - Type of industry respondent is employed in. Values are represented as short random character strings.
        * employment_occupation - Type of occupation of respondent. Values are represented as short random character strings.
* What is the target variable?
    * There are two target variables in this dataset:
        * h1n1_vaccine - Whether respondent received H1N1 flu vaccine.
        * seasonal_vaccine - Whether respondent received seasonal flu vaccine.
    * This project will focus on the h1n1_vaccine target, specifically.
* What are the properties of the variables you intend to use?
    * We intend to use all of the available variables, at least until any given variable is shown to be statistically insignificant.
    * Most of the variables have numeric data types and will likely be left as-is. The variables which do not have numeric data types will likely be transformed using ordinal encoding or one hot encoding, as appropriate.

***

The above descriptions of the features in this dataset were taken from DrivenData at the following URL:

https://www.drivendata.org/competitions/66/flu-shot-learning/page/211/#features_list

The 10 HHS geographic classification regions:

![regional-offices.png](images/regional-offices.png)

The above image was taken from the U.S. Department of Health and Human Services at the following URL:

https://www.hhs.gov/about/agencies/iea/regional-offices/index.html

A brief word on MSA:

"Standard delineations of metropolitan areas were first issued in 1949 by the then Bureau of the Budget (predecessor of OMB), under the designation "standard metropolitan area" (SMA). The term was changed to "standard metropolitan statistical area" (SMSA) in 1959, and to "metropolitan statistical area" (MSA) in 1983. The term "metropolitan area" (MA) was adopted in 1990 and referred collectively to metropolitan statistical areas (MSAs), consolidated metropolitan statistical areas (CMSAs), and primary metropolitan statistical areas (PMSAs). The term "core based statistical area" (CBSA) became effective in 2000 and refers collectively to metropolitan and micropolitan statistical areas."

The above excerpt on MSA was taken from the United States Census Bureau at the following URL:

https://www.census.gov/programs-surveys/metro-micro/about.html

Let's go ahead and import the standard packages we will use for this project.

In [1]:
# Import standard packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

Let's go ahead and read in the files.

In [2]:
# Read in the features of the training set.
features = pd.read_csv('./data/features.csv')

# Quickly test the type of data in the dataframe.
display(features.head())
display(features.tail())

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
26702          26702           2.0             0.0                        0.0   
26703          26703           1.0             2.0                        0.0   
26704          26704           2.0             2.0                        0.0   
26705          26705           1.0             1.0                        0.0   
26706          26706           0.0             0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
26702                   1.0                   0.0                    0.0   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   1.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
26702                          0.0                      1.0   
26703                          0.0                      0.0   
26704                          1.0                      0.0   
26705                          0.0                      0.0   
26706                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
26702                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26703                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26704                    1.0  ...                        NaN     Not Married   
26705                    NaN  ...  <= $75,000, Above Poverty         Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   

       rent_or_own   employment_status  hhs_geo_region  \
26702          Own  Not in Labor Force        qufhixun   
26703         Rent            Employed        lzgpxyit   
26704          Own                 NaN        lzgpxyit   
26705         Rent            Employed        lrircsnp   
26706          Own  Not in Labor Force        mlyzmhmf   

                     census_msa  household_adults  household_children  \
26702                   Non-MSA               0.0                 0.0   
26703       MSA, Principle City               1.0                 0.0   
26704  MSA, Not Principle  City               0.0                 0.0   
26705                   Non-MSA               1.0                 0.0   
26706       MSA, Principle City               1.0                 0.0   

       employment_industry  employment_occupation  
26702                  NaN                    NaN  
26703             fcxhlnwr               cmhcxjea  
26704                  NaN                    NaN  
26705             fcxhlnwr               haliazsg  
26706                  NaN                    NaN  

[5 rows x 36 columns]

As expected, the features dataframe has 36 columns, which corresponds correctly to our number of features, plus the respondent_id.<br>One thing worth noting here is that the dataframe indices and the respondent_id correspond perfectly, so we will likely set the index of the columns to respondent_id.

Let's go ahead and read in the rest of the files and confirm whether or not they load in as expected.

In [3]:
# Read in the labels of the training set.
labels = pd.read_csv('./data/labels.csv')

# Quickly test the type of data in the dataframe.
display(labels.head())
display(labels.tail())

respondent_id  h1n1_vaccine  seasonal_vaccine
0              0             0                 0
1              1             0                 1
2              2             0                 0
3              3             0                 1
4              4             0                 0

respondent_id  h1n1_vaccine  seasonal_vaccine
26702          26702             0                 0
26703          26703             0                 0
26704          26704             0                 1
26705          26705             0                 0
26706          26706             0                 0

As expected, the labels dataframe has 3 columns, which corresponds correctly to our number of labels, plus the respondent_id.<br>However, the respondent_id and the dataframe indices again correspond perfectly, so we will likely set the index of the columns to respondent_id.<br>

Now that all of the files have been read in correctly, let's concatenate them into a dataframe so that we can execute our own train test splits on them and use them as a complete dataframe in later pipelines.

In [4]:
# Concatenate the dataframes.
df = pd.concat([features, labels], axis=1)

# Quickly test the type of data in the concatenated dataframe.
display(df.head())
display(df.tail())

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...   employment_status  hhs_geo_region  \
0                    1.0  ...  Not in Labor Force        oxchjgsf   
1                    1.0  ...            Employed        bhuqouqj   
2                    0.0  ...            Employed        qufhixun   
3                    0.0  ...  Not in Labor Force        lrircsnp   
4                    1.0  ...            Employed        qufhixun   

                 census_msa  household_adults  household_children  \
0                   Non-MSA               0.0                 0.0   
1  MSA, Not Principle  City               0.0                 0.0   
2  MSA, Not Principle  City               2.0                 0.0   
3       MSA, Principle City               0.0                 0.0   
4  MSA, Not Principle  City               1.0                 0.0   

   employment_industry  employment_occupation  respondent_id  h1n1_vaccine  \
0                  NaN                    NaN              0             0   
1             pxcmvdjn               xgwztkwe              1             0   
2             rucpziij               xtkaffoo              2             0   
3                  NaN                    NaN              3             0   
4             wxleyezf               emcorrxb              4             0   

   seasonal_vaccine  
0                 0  
1                 1  
2                 0  
3                 1  
4                 0  

[5 rows x 39 columns]

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
26702          26702           2.0             0.0                        0.0   
26703          26703           1.0             2.0                        0.0   
26704          26704           2.0             2.0                        0.0   
26705          26705           1.0             1.0                        0.0   
26706          26706           0.0             0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
26702                   1.0                   0.0                    0.0   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   1.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
26702                          0.0                      1.0   
26703                          0.0                      0.0   
26704                          1.0                      0.0   
26705                          0.0                      0.0   
26706                          0.0                      0.0   

       behavioral_touch_face  ...   employment_status  hhs_geo_region  \
26702                    0.0  ...  Not in Labor Force        qufhixun   
26703                    0.0  ...            Employed        lzgpxyit   
26704                    1.0  ...                 NaN        lzgpxyit   
26705                    NaN  ...            Employed        lrircsnp   
26706                    0.0  ...  Not in Labor Force        mlyzmhmf   

                     census_msa  household_adults  household_children  \
26702                   Non-MSA               0.0                 0.0   
26703       MSA, Principle City               1.0                 0.0   
26704  MSA, Not Principle  City               0.0                 0.0   
26705                   Non-MSA               1.0                 0.0   
26706       MSA, Principle City               1.0                 0.0   

       employment_industry  employment_occupation  respondent_id  \
26702                  NaN                    NaN          26702   
26703             fcxhlnwr               cmhcxjea          26703   
26704                  NaN                    NaN          26704   
26705             fcxhlnwr               haliazsg          26705   
26706                  NaN                    NaN          26706   

       h1n1_vaccine  seasonal_vaccine  
26702             0                 0  
26703             0                 0  
26704             0                 1  
26705             0                 0  
26706             0                 0  

[5 rows x 39 columns]

Perfect. Now we have a single dataframe. However, our respondent_id column is in the dataframe twice now, so we will have to drop one of them. Fortunately, the dataframe index is actually the same as the respondent_id columns, so we can actually drop them both.

In [5]:
# Drop both of the respondent_id columns and save the changes to the df object.
df.drop(columns='respondent_id', inplace=True)

# Quickly test the type of data in the concatenated dataframe.
display(df.head())
display(df.tail())

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0           1.0             0.0                        0.0   
1           3.0             2.0                        0.0   
2           1.0             1.0                        0.0   
3           1.0             1.0                        0.0   
4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  doctor_recc_h1n1  ...  rent_or_own  \
0                    1.0               0.0  ...          Own   
1                    1.0               0.0  ...         Rent   
2                    0.0               NaN  ...          Own   
3                    0.0               0.0  ...         Rent   
4                    1.0               0.0  ...          Own   

    employment_status  hhs_geo_region                census_msa  \
0  Not in Labor Force        oxchjgsf                   Non-MSA   
1            Employed        bhuqouqj  MSA, Not Principle  City   
2            Employed        qufhixun  MSA, Not Principle  City   
3  Not in Labor Force        lrircsnp       MSA, Principle City   
4            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  h1n1_vaccine  seasonal_vaccine  
0                    NaN             0                 0  
1               xgwztkwe             0                 1  
2               xtkaffoo             0                 0  
3                    NaN             0                 1  
4               emcorrxb             0                 0  

[5 rows x 37 columns]

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
26702           2.0             0.0                        0.0   
26703           1.0             2.0                        0.0   
26704           2.0             2.0                        0.0   
26705           1.0             1.0                        0.0   
26706           0.0             0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
26702                   1.0                   0.0                    0.0   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   1.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
26702                          0.0                      1.0   
26703                          0.0                      0.0   
26704                          1.0                      0.0   
26705                          0.0                      0.0   
26706                          0.0                      0.0   

       behavioral_touch_face  doctor_recc_h1n1  ...  rent_or_own  \
26702                    0.0               0.0  ...          Own   
26703                    0.0               1.0  ...         Rent   
26704                    1.0               0.0  ...          Own   
26705                    NaN               0.0  ...         Rent   
26706                    0.0               0.0  ...          Own   

        employment_status  hhs_geo_region                census_msa  \
26702  Not in Labor Force        qufhixun                   Non-MSA   
26703            Employed        lzgpxyit       MSA, Principle City   
26704                 NaN        lzgpxyit  MSA, Not Principle  City   
26705            Employed        lrircsnp                   Non-MSA   
26706  Not in Labor Force        mlyzmhmf       MSA, Principle City   

       household_adults  household_children  employment_industry  \
26702               0.0                 0.0                  NaN   
26703               1.0                 0.0             fcxhlnwr   
26704               0.0                 0.0                  NaN   
26705               1.0                 0.0             fcxhlnwr   
26706               1.0                 0.0                  NaN   

       employment_occupation  h1n1_vaccine  seasonal_vaccine  
26702                    NaN             0                 0  
26703               cmhcxjea             0                 0  
26704                    NaN             0                 1  
26705               haliazsg             0                 0  
26706                    NaN             0                 0  

[5 rows x 37 columns]

Excellent. Now we have a complete dataframe with all of our features and our target, with no duplicate columns, and our index is set to the respondent_id.

Before we start making any changes to the complete dataframe, let's make a copy of it in case something gets corrupted.

In [6]:
# Assign a copy dataframe variable and leave it alone for use in a backup situation.
df_copy = df.copy()

# Quickly test the type of data in the concatenated dataframe.
display(df_copy.head())
display(df_copy.tail())

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0           1.0             0.0                        0.0   
1           3.0             2.0                        0.0   
2           1.0             1.0                        0.0   
3           1.0             1.0                        0.0   
4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  doctor_recc_h1n1  ...  rent_or_own  \
0                    1.0               0.0  ...          Own   
1                    1.0               0.0  ...         Rent   
2                    0.0               NaN  ...          Own   
3                    0.0               0.0  ...         Rent   
4                    1.0               0.0  ...          Own   

    employment_status  hhs_geo_region                census_msa  \
0  Not in Labor Force        oxchjgsf                   Non-MSA   
1            Employed        bhuqouqj  MSA, Not Principle  City   
2            Employed        qufhixun  MSA, Not Principle  City   
3  Not in Labor Force        lrircsnp       MSA, Principle City   
4            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  h1n1_vaccine  seasonal_vaccine  
0                    NaN             0                 0  
1               xgwztkwe             0                 1  
2               xtkaffoo             0                 0  
3                    NaN             0                 1  
4               emcorrxb             0                 0  

[5 rows x 37 columns]

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
26702           2.0             0.0                        0.0   
26703           1.0             2.0                        0.0   
26704           2.0             2.0                        0.0   
26705           1.0             1.0                        0.0   
26706           0.0             0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
26702                   1.0                   0.0                    0.0   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   1.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
26702                          0.0                      1.0   
26703                          0.0                      0.0   
26704                          1.0                      0.0   
26705                          0.0                      0.0   
26706                          0.0                      0.0   

       behavioral_touch_face  doctor_recc_h1n1  ...  rent_or_own  \
26702                    0.0               0.0  ...          Own   
26703                    0.0               1.0  ...         Rent   
26704                    1.0               0.0  ...          Own   
26705                    NaN               0.0  ...         Rent   
26706                    0.0               0.0  ...          Own   

        employment_status  hhs_geo_region                census_msa  \
26702  Not in Labor Force        qufhixun                   Non-MSA   
26703            Employed        lzgpxyit       MSA, Principle City   
26704                 NaN        lzgpxyit  MSA, Not Principle  City   
26705            Employed        lrircsnp                   Non-MSA   
26706  Not in Labor Force        mlyzmhmf       MSA, Principle City   

       household_adults  household_children  employment_industry  \
26702               0.0                 0.0                  NaN   
26703               1.0                 0.0             fcxhlnwr   
26704               0.0                 0.0                  NaN   
26705               1.0                 0.0             fcxhlnwr   
26706               1.0                 0.0                  NaN   

       employment_occupation  h1n1_vaccine  seasonal_vaccine  
26702                    NaN             0                 0  
26703               cmhcxjea             0                 0  
26704                    NaN             0                 1  
26705               haliazsg             0                 0  
26706                    NaN             0                 0  

[5 rows x 37 columns]

Now that we have a df object to use, let's conduct our preliminary data exploration.

In [7]:
# Let's check out the df info to verify how many features we have, what they are,
# how many nulls we have, and the data types of our features.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 37 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   h1n1_concern                 26615 non-null  float64
 1   h1n1_knowledge               26591 non-null  float64
 2   behavioral_antiviral_meds    26636 non-null  float64
 3   behavioral_avoidance         26499 non-null  float64
 4   behavioral_face_mask         26688 non-null  float64
 5   behavioral_wash_hands        26665 non-null  float64
 6   behavioral_large_gatherings  26620 non-null  float64
 7   behavioral_outside_home      26625 non-null  float64
 8   behavioral_touch_face        26579 non-null  float64
 9   doctor_recc_h1n1             24547 non-null  float64
 10  doctor_recc_seasonal         24547 non-null  float64
 11  chronic_med_condition        25736 non-null  float64
 12  child_under_6_months         25887 non-null  float64
 13  health_worker   

Again, we can see that we have the 35 expected features plus our targets, which we will split with our train test split.<br>The first thing that stands out here is that there are a significant number of nulls for the features 'health_insurance', 'employment_industry', and 'employment_occupation'. We need to take a look at these features to see how best to deal with those nulls and whether or not those features will be useable in our models. We will deal with these forms of data preparation shortly.

Let's explore the shape of our dataframe.

In [8]:
# Let's observe how many rows and columns are in our dataframe.
df.shape

(26707, 37)

Based on how shape compares to the indices of the dataframe, we can confidently say that every index is unique and the distribution of indices across the dataframe is uniform.

Let's explore the size of our dataframe.

In [9]:
# Check the number of rows by columns there are
df.size

988159

It looks like there are just under 1 million individual intersections of row and column in this dataset. Hopefully this will give us enough datapoints to create some robust machine learning models.

Let's explore the number of dimensions that our dataframe has.

In [10]:
# Check the dimensions
df.ndim

2

Nothing surprising here; it is simply rows by columns, as expected.

Let's explore the numerical description of our dataframe.

In [11]:
# Split the descriptions into two displays to see all features.
display(df.iloc[:, :12].describe())
display(df.iloc[:, 12:].describe())

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
count  26615.000000    26591.000000               26636.000000   
mean       1.618486        1.262532                   0.048844   
std        0.910311        0.618149                   0.215545   
min        0.000000        0.000000                   0.000000   
25%        1.000000        1.000000                   0.000000   
50%        2.000000        1.000000                   0.000000   
75%        2.000000        2.000000                   0.000000   
max        3.000000        2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          26499.000000          26688.000000           26665.000000   
mean               0.725612              0.068982               0.825614   
std                0.446214              0.253429               0.379448   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  26620.00000             26625.000000   
mean                       0.35864                 0.337315   
std                        0.47961                 0.472802   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  doctor_recc_h1n1  doctor_recc_seasonal  \
count           26579.000000      24547.000000          24547.000000   
mean                0.677264          0.220312              0.329735   
std                 0.467531          0.414466              0.470126   
min                 0.000000          0.000000              0.000000   
25%                 0.000000          0.000000              0.000000   
50%                 1.000000          0.000000              0.000000   
75%                 1.000000          0.000000              1.000000   
max                 1.000000          1.000000              1.000000   

       chronic_med_condition  
count           25736.000000  
mean                0.283261  
std                 0.450591  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 1.000000  
max                 1.000000

child_under_6_months  health_worker  health_insurance  \
count          25887.000000   25903.000000       14433.00000   
mean               0.082590       0.111918           0.87972   
std                0.275266       0.315271           0.32530   
min                0.000000       0.000000           0.00000   
25%                0.000000       0.000000           1.00000   
50%                0.000000       0.000000           1.00000   
75%                0.000000       0.000000           1.00000   
max                1.000000       1.000000           1.00000   

       opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
count                 26316.000000       26319.000000   
mean                      3.850623           2.342566   
std                       1.007436           1.285539   
min                       1.000000           1.000000   
25%                       3.000000           1.000000   
50%                       4.000000           2.000000   
75%                       5.000000           4.000000   
max                       5.000000           5.000000   

       opinion_h1n1_sick_from_vacc  opinion_seas_vacc_effective  \
count                 26312.000000                 26245.000000   
mean                      2.357670                     4.025986   
std                       1.362766                     1.086565   
min                       1.000000                     1.000000   
25%                       1.000000                     4.000000   
50%                       2.000000                     4.000000   
75%                       4.000000                     5.000000   
max                       5.000000                     5.000000   

       opinion_seas_risk  opinion_seas_sick_from_vacc  household_adults  \
count       26193.000000                 26170.000000      26458.000000   
mean            2.719162                     2.118112          0.886499   
std             1.385055                     1.332950          0.753422   
min             1.000000                     1.000000          0.000000   
25%             2.000000                     1.000000          0.000000   
50%             2.000000                     2.000000          1.000000   
75%             4.000000                     4.000000          1.000000   
max             5.000000                     5.000000          3.000000   

       household_children  h1n1_vaccine  seasonal_vaccine  
count        26458.000000  26707.000000      26707.000000  
mean             0.534583      0.212454          0.465608  
std              0.928173      0.409052          0.498825  
min              0.000000      0.000000          0.000000  
25%              0.000000      0.000000          0.000000  
50%              0.000000      0.000000          0.000000  
75%              1.000000      0.000000          1.000000  
max              3.000000      1.000000          1.000000

The things that stick out here as being potentially noteworthy are these:
* First, the mean of behavioral_antiviral_meds is incredibly low.
* Second, the mean of behavioral_face_mask is incredibly low.
* Third, the means and standard distributions of the opinion_h1n1_vacc_effective, opinion_h1n1_risk, and opinion_h1n1_sick_from_vacc are quite similar to the means and standard distributions of opinion_seas_vacc_effective, opinion_seas_risk, and opinion_seas_sick_from_vacc, respectively.
* Fourth, despite having quite a high mean for opinion_h1n1_vacc_effective and moderate means for opinion_h1n1_risk and opinion_h1n1_sick_from_vacc, there is a fairly low mean for h1n1_vaccine.
* Finally, despite the similarities between the h1n1 opinion features and the seasonal flu opinion features, the mean for seasonal_vaccine is much higher than the mean for h1n1_vaccine.

Let's take a look at a random sample from our dataframe.

In [12]:
display(df.iloc[:, :18].sample(random_state=42))
display(df.iloc[:, 18:].sample(random_state=42))

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
15772           2.0             1.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
15772                   1.0                   1.0                    1.0   

       behavioral_large_gatherings  behavioral_outside_home  \
15772                          1.0                      1.0   

       behavioral_touch_face  doctor_recc_h1n1  doctor_recc_seasonal  \
15772                    1.0               0.0                   0.0   

       chronic_med_condition  child_under_6_months  health_worker  \
15772                    NaN                   NaN            NaN   

       health_insurance  opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
15772               NaN                          NaN                NaN   

       opinion_h1n1_sick_from_vacc  
15772                          NaN

opinion_seas_vacc_effective  opinion_seas_risk  \
15772                          NaN                NaN   

       opinion_seas_sick_from_vacc      age_group education   race     sex  \
15772                          NaN  18 - 34 Years       NaN  White  Female   

      income_poverty marital_status rent_or_own employment_status  \
15772            NaN            NaN         NaN               NaN   

      hhs_geo_region           census_msa  household_adults  \
15772       atmpeygn  MSA, Principle City               NaN   

       household_children employment_industry employment_occupation  \
15772                 NaN                 NaN                   NaN   

       h1n1_vaccine  seasonal_vaccine  
15772             0                 0

We can say the following things about this particular respondent:
* They were **somewhat concerned** about the H1N1 flu.
* They had **a little knowledge** about H1N1 flu.
* They had **not** taken antiviral medications.
* They **had** avoided close contact with others with flu-like symptoms.
* They **had** bought a face mask.
* They **had** frequently washed their hands or use hand sanitizer.
* They **had** reduced their time at large gatherings.
* They **had** reduced their contact with people outside of their own household.
* They **had** avoided touching their eyes, their nose, or their mouth.
* The H1N1 flu vaccine was **not** recommended to them by a doctor.
* The seasonal flu vaccine was **not** recommended to them by a doctor.

This particular respondent did not answer many of the other questions, but we do also know the following:
* They are between the ages of **18** and **34**.
* They are **white**.
* They are **female**.
* They did **not** get the **H1N1** flu vaccine **or** the **seasonal** flu vaccine.

One thing that may be interesting about the data itself is that, while the dataset as a whole does not have a large number of nulls (apart from the features we noticed earlier), this particular respondent had null entries for many features. This may indicate an overlap of null value counts between features in the entries for specific respondents, so we may be able to drop the remaining nulls after dealing with our large null count features.

Let's take a look at how many nulls there are by feature in the dataframe.

In [13]:
df.isnull().sum()

h1n1_concern                      92
h1n1_knowledge                   116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_h1n1                2160
doctor_recc_seasonal            2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_h1n1_vacc_effective      391
opinion_h1n1_risk                388
opinion_h1n1_sick_from_vacc      395
opinion_seas_vacc_effective      462
opinion_seas_risk                514
opinion_seas_sick_from_vacc      537
age_group                          0
education                       1407
race                               0
sex                                0
income_poverty                  4423
marital_status                  1408
r

Some interesting things to note here are these:
* The number of null counts for doctor_recc_h1n1 and the number of null counts for doctor_recc_seasonal are exactly the same.
* The number of null counts for the features health_insurance, employment_industry, and employment_occupation are all quite high, which is consistent with our findings from earlier.
* Given a sample size of 26,707, the only feature other than those three with a null count higher than 10% of the sample size (2,670) is income_poverty, which has a null count of 4,423. Depending on how many respondents submitted null responses for multiple features, it may be possible to simply drop all of the nulls from the dataframe after dealing with the three large null count features from earlier, **with the exception of income_poverty**. We will either have to find a way to impute some values for the feature or drop it, as well, depending on the distribution of its value counts.
* The only features with no null counts at all are age_group, race, sex, hhs_geo_region, census_msa, h1n1_vaccine, and seasonal_vaccine. You may recall that these were the only features our sample respondent answered, after docc_recc_seasonal. This may tell us that answers to these questions were mandatory, or that any survey response which was missing a value in any of these questions was not included in the dataset.

In [14]:
df.nunique()

h1n1_concern                    4
h1n1_knowledge                  3
behavioral_antiviral_meds       2
behavioral_avoidance            2
behavioral_face_mask            2
behavioral_wash_hands           2
behavioral_large_gatherings     2
behavioral_outside_home         2
behavioral_touch_face           2
doctor_recc_h1n1                2
doctor_recc_seasonal            2
chronic_med_condition           2
child_under_6_months            2
health_worker                   2
health_insurance                2
opinion_h1n1_vacc_effective     5
opinion_h1n1_risk               5
opinion_h1n1_sick_from_vacc     5
opinion_seas_vacc_effective     5
opinion_seas_risk               5
opinion_seas_sick_from_vacc     5
age_group                       5
education                       4
race                            4
sex                             2
income_poverty                  3
marital_status                  2
rent_or_own                     2
employment_status               3
hhs_geo_region

Looking at the above collection of unique values in each feature, we can see that all of our numeric features have the expected number of values, as laid out by the feature descriptions above. However, we don't yet know what the various value counts for the object type columns are, so let's take a look at those briefly. In addition to simply understanding what sorts of answers were available on the survey, this will help us determine what sort of transformation would be appropriate for each feature.

# PICK UP AGAIN HERE

In [15]:
df.columns

Index(['h1n1_concern', 'h1n1_knowledge', 'behavioral_antiviral_meds',
       'behavioral_avoidance', 'behavioral_face_mask', 'behavioral_wash_hands',
       'behavioral_large_gatherings', 'behavioral_outside_home',
       'behavioral_touch_face', 'doctor_recc_h1n1', 'doctor_recc_seasonal',
       'chronic_med_condition', 'child_under_6_months', 'health_worker',
       'health_insurance', 'opinion_h1n1_vacc_effective', 'opinion_h1n1_risk',
       'opinion_h1n1_sick_from_vacc', 'opinion_seas_vacc_effective',
       'opinion_seas_risk', 'opinion_seas_sick_from_vacc', 'age_group',
       'education', 'race', 'sex', 'income_poverty', 'marital_status',
       'rent_or_own', 'employment_status', 'hhs_geo_region', 'census_msa',
       'household_adults', 'household_children', 'employment_industry',
       'employment_occupation', 'h1n1_vaccine', 'seasonal_vaccine'],
      dtype='object')

In [16]:
df.dtypes.index

Index(['h1n1_concern', 'h1n1_knowledge', 'behavioral_antiviral_meds',
       'behavioral_avoidance', 'behavioral_face_mask', 'behavioral_wash_hands',
       'behavioral_large_gatherings', 'behavioral_outside_home',
       'behavioral_touch_face', 'doctor_recc_h1n1', 'doctor_recc_seasonal',
       'chronic_med_condition', 'child_under_6_months', 'health_worker',
       'health_insurance', 'opinion_h1n1_vacc_effective', 'opinion_h1n1_risk',
       'opinion_h1n1_sick_from_vacc', 'opinion_seas_vacc_effective',
       'opinion_seas_risk', 'opinion_seas_sick_from_vacc', 'age_group',
       'education', 'race', 'sex', 'income_poverty', 'marital_status',
       'rent_or_own', 'employment_status', 'hhs_geo_region', 'census_msa',
       'household_adults', 'household_children', 'employment_industry',
       'employment_occupation', 'h1n1_vaccine', 'seasonal_vaccine'],
      dtype='object')

# END HERE

## Data Preparation

Desscribe and justify the process for preparing the data for analysis.

***

Questions to consider:

* Were there variables you dropped or created?
* How did you address missing values or outliers?
* Why are these choices appropriate given the data and the business problem?

***

In [ ]:
# Examine the value counts of the health insurance feature to observe its distribution.
df['health_insurance'].value_counts()

So it looks like all of our values are either 1 or 0, again as expected. In terms of what we could do to rectify the null counts in this column, we could potentially impute the mean or the median, but both of these approaches will likely fudge the data. If we impute the mean, we change the distribution of the entire column from binary to something else, so we can't do that. If we impute the median, we drastically sway an already imbalanced distribution even more in favor of the majority, so we can't really do that, either. With respect to this column missing as many values as it does, the best approach for this feature may simply be to drop it altogether.

In [ ]:
# Drop the health insurance feature from the dataset, as it is missing too many values to use.
df.drop(columns='health_insurance', inplace=True)

# Quickly test the type of data in the concatenated dataframe.
display(df.head())
display(df.tail())

Now that we can see that the null count of income_poverty is also somewhat high, let's take a look at the value counts distribution of that feature.

In [ ]:
# Explore the various values for income_poverty and how times each value occurs in the dataframe.
df['income_poverty'].value_counts()

Well, unfortunately, we may be in a similar pickle to the one we were in with the big three, as imputing 4,423 entries for any of these values will have a significant impact on the distribution of this feature. We'll go ahead and drop this feature from our modeling process.

## Data Modeling

Describe and justify the process for analyzing or modeling the data.

***

Questions to consider:

* How did you analyze or model the data?
* How did you iterate on your initial approach to make it better?
* Why are these choices appropriate given the data and the business problem?

***

In [ ]:
# Here you run your code to model the data

##  Evaluation

Evaluate how well your work solves the stated business problem.

***

Questions to consider:

* How do you interpret the results?
* How well does your model fir your data?
* How much better is this than your baseline model?
* How confident are you that your results would generalize beyond the data you have?
* How confident are you that this model would benefit the business if put into use?

***

## Conclusions

Provide your conclusions about the work you've done, including any limitations or next steps.

***

Questions to consider:

* What would you recommend the business do as a result of this work?
* What are some reasons why your analysis might not fully solve the business problem?
* What else could you do  in the future to improve this project?

***